pkg/service/iam/handler_account.go

```diff
@@ -0,0 +1,387 @@
+	if err != nil {
+		return nil, err
+	}
+
+	reply := &pbiam.GetPasswordResetResponse{
+		ResetId: pbutil.ToProtoString(reset_id),
+		UserId:  pbutil.ToProtoString(reset_info.ResetId),
+	}
+
+	return reply, nil
+}
+func (p *Server) ValidateUserPassword(ctx context.Context, req *pbiam.ValidateUserPasswordRequest) (*pbiam.ValidateUserPasswordResponse, error) {
+	var email = req.GetEmail()
+	var userInfo models.User
+
+	query := pi.Global().DB(ctx).
+		Select(models.UserColumns...).
+		From(models.UserTableName).Limit(1).
+		Where(db.Eq(models.ColumnEmail, email))
+	err := query.LoadOne(&userInfo)
+	if err != nil {
+		return nil, fmt.Errorf("user(%q) not fount", email)
+	}
+
+	reply := &pbiam.ValidateUserPasswordResponse{
+		Validated: userInfo.Password == req.GetPassword(),
+	}
+
+	return reply, nil
+}
+
+func (p *Server) CreateGroup(ctx context.Context, req *pbiam.CreateGroupRequest) (*pbiam.CreateGroupResponse, error) {
+	var newGroup = models.NewGroup(
+		req.GetName().GetValue(),
+		req.GetDescription().GetValue(),
+	)
+
+	_, err := pi.Global().DB(ctx).
+		InsertInto(models.GroupTableName).
+		Columns(models.GroupColumns...).
+		Record(newGroup).
+		Exec()
+	if err != nil {
+		return nil, gerr.NewWithDetail(ctx, gerr.Internal, err, gerr.ErrorCreateResourcesFailed)
+	}
+
+	reply := &pbiam.CreateGroupResponse{
+		GroupId: pbutil.ToProtoString(newGroup.GroupId),
+	}
+
+	return reply, nil
+}
+
+func (p *Server) ModifyGroup(ctx context.Context, req *pbiam.ModifyGroupRequest) (*pbiam.ModifyGroupResponse, error) {
+	group_id := req.GetGroupId().GetValue()
+	if group_id == "" {
+		return &pbiam.ModifyGroupResponse{}, nil
+	}
+
+	var attributes = make(map[string]interface{})
+	if req.Name != nil {
+		attributes[models.ColumnName] = req.Name.GetValue()
+	}
+	if req.Description != nil {
+		attributes[models.ColumnDescription] = req.Description.GetValue()
+	}
+
+	if len(attributes) == 0 {
+		return &pbiam.ModifyGroupResponse{}, nil
+	}
+
+	_, err := pi.Global().DB(ctx).
+		Update(models.GroupTableName).SetMap(attributes).
+		Where(db.Eq(models.ColumnGroupId, group_id)).
+		Exec()
+	if err != nil {
+		return nil, gerr.NewWithDetail(ctx, gerr.Internal, err, gerr.ErrorDeleteResourcesFailed)
+	}
+
+	reply := &pbiam.ModifyGroupResponse{
+		GroupId: pbutil.ToProtoString(group_id),
+	}
+
+	return reply, nil
+}
+func (p *Server) DeleteGroups(ctx context.Context, req *pbiam.DeleteGroupsRequest) (*pbiam.DeleteGroupsResponse, error) {
+	_, err := pi.Global().DB(ctx).
+		DeleteFrom(models.GroupTableName).
+		Where(db.Eq(models.ColumnGroupId, req.GroupId)).
+		Exec()
+	if err != nil {
+		return nil, gerr.NewWithDetail(ctx, gerr.Internal, err, gerr.ErrorDeleteResourcesFailed)
+	}
+
+	reply := &pbiam.DeleteGroupsResponse{
+		GroupId: req.GroupId,
+	}
+
+	return reply, nil
+}
+
+func (p *Server) JoinGroup(ctx context.Context, req *pbiam.JoinGroupRequest) (*pbiam.JoinGroupResponse, error) {
+	if len(req.GroupId) != 1 || len(req.UserId) != 1 {
```

Comment:
Loop user_id and group_id to generate the record to be inserted. Check if the record exists before insertion and skip if it exists.
`len(req.GroupId) != 0 && len(req.UserId) != 0`